src/Swarm/Language/Typecheck.hs

```diff
@@ -206,14 +207,20 @@ skolemize (Forall xs uty) = do
 
 -- | 'generalize' is the opposite of 'instantiate': add a 'Forall'
 --   which closes over all free type and unification variables.
+--
+--   Pick nice type variable names instead of reusing whatever fresh
+--   names happened to be used for the free variables.
 generalize :: UType -> Infer UPolytype
 generalize uty = do
   uty' <- applyBindings uty
   ctx <- ask
   tmfvs <- freeVars uty'
   ctxfvs <- freeVars ctx
   let fvs = S.toList $ tmfvs \\ ctxfvs
-      xs = map (mkVarName "a") fvs
+      alphabet = ['a' .. 'z']
+      xs =
+        take (length fvs) $
```

Comment:
Is this `take (length fvs) $` necessary?  The `zip` on line 224 would to the proper list truncation regardless?

Reply:
You know, I thought the same thing, and so I left out this call to `take` when I initially wrote the code.  Then my computer almost fell over when the test suite started to use all available memory. :smile:   The problem is that `xs` is *also* used as the first argument to `Forall`, so if `xs` is infinite...

I think I can refactor it to be more readable/obvious though...